storage/src/adb/keyless.rs

```diff
@@ -0,0 +1,863 @@
+//! its location.
+
+use crate::{
+    adb::Error,
+    journal::{
+        fixed::{Config as FConfig, Journal as FJournal},
+        variable::{Config as VConfig, Journal as VJournal},
+    },
+    mmr::{
+        hasher::Standard,
+        iterator::{leaf_num_to_pos, leaf_pos_to_num},
+        journaled::{Config as MmrConfig, Mmr},
+        verification::Proof,
+    },
+    store::operation::Keyless as Operation,
+};
+use commonware_codec::{Codec, Encode as _};
+use commonware_cryptography::Hasher as CHasher;
+use commonware_runtime::{buffer::PoolRef, Clock, Metrics, Storage, ThreadPool};
+use futures::{future::TryFutureExt, try_join};
+use std::num::{NonZeroU64, NonZeroUsize};
+use tracing::warn;
+
+/// Configuration for a [Keyless] authenticated db.
+#[derive(Clone)]
+pub struct Config<C> {
+    /// The name of the [Storage] partition used for the MMR's backing journal.
+    pub mmr_journal_partition: String,
+
+    /// The items per blob configuration value used by the MMR journal.
+    pub mmr_items_per_blob: NonZeroU64,
+
+    /// The size of the write buffer to use for each blob in the MMR journal.
+    pub mmr_write_buffer: NonZeroUsize,
+
+    /// The name of the [Storage] partition used for the MMR's metadata.
+    pub mmr_metadata_partition: String,
+
+    /// The name of the [Storage] partition used to persist the (pruned) values.
+    pub values_journal_partition: String,
+
+    /// The size of the write buffer to use with the values journal.
+    pub values_write_buffer: NonZeroUsize,
+
+    /// Optional compression level (using `zstd`) to apply to values data before storing.
+    pub values_compression: Option<u8>,
+
+    /// The codec configuration to use for encoding and decoding values.
+    pub values_codec_config: C,
+
+    /// The max number of values to put in each section of the values journal.
+    pub values_items_per_section: NonZeroU64,
+
+    /// The name of the [Storage] partition used for the location map.
+    pub locations_journal_partition: String,
+
+    /// The number of items to put in each blob in the location map.
+    pub locations_items_per_blob: NonZeroU64,
+
+    /// The size of the write buffer to use with the locations journal.
+    pub locations_write_buffer: NonZeroUsize,
+
+    /// An optional thread pool to use for parallelizing batch operations.
+    pub thread_pool: Option<ThreadPool>,
+
+    /// The buffer pool to use for caching data.
+    pub buffer_pool: PoolRef,
+}
+
+/// A keyless ADB for variable length data.
+pub struct Keyless<E: Storage + Clock + Metrics, V: Codec, H: CHasher> {
+    /// An MMR over digests of the operations applied to the db.
+    ///
+    /// # Invariant
+    ///
+    /// The number of leaves in this MMR always equals the number of operations in the unpruned
+    /// `locations` journal.
+    mmr: Mmr<E, H>,
+
+    /// A (pruned) journal of all values ever appended to the db.
```

Comment:
should you also note `(pruned)` here and for `mmr` like you did for `values`?

I think it may be cleaner to just remove that note (as we assume a number of things will be pruned.

Reply:
agreed done